relayer/relayer.go

```diff
@@ -158,63 +164,102 @@ func NewRelayer(
-
-	// The following cases are treated as successful:
-	// 1) The database contains the latest processed block data for the chain
-	//    - In this case, we parse the block height and process warp logs from that height to the current block
-	// 2) The database has been configured for the chain, but does not contain the latest processed block data
-	//    - In this case, we save the current block height in the database, but do not process any historical warp logs
-	if err == nil {
-		// If the database contains the latest processed block data, then back-process all warp messages from that block to the latest block
-		// Note that the retrieved latest processed block may have already been partially (or fully) processed by the relayer on a previous run. When
-		// processing a warp message in real time, which is when we update the latest processed block in the database, we have no way of knowing
-		// if that is the last warp message in the block
-		latestProcessedBlock, success := new(big.Int).SetString(string(latestProcessedBlockData), 10)
-		if !success {
-			r.logger.Error("failed to convert latest block to big.Int", zap.Error(err))
-			return err
-		}
-
-		err = sub.ProcessFromHeight(latestProcessedBlock)
-		if err != nil {
+	if errors.Is(err, database.ErrChainNotFound) || errors.Is(err, database.ErrKeyNotFound) {
+		// The database does not contain the latest processed block data for the chain, so use the configured StartBlockHeight instead
+		if startBlockHeight == 0 {
 			r.logger.Warn(
-				"Encountered an error when processing historical blocks. Continuing to normal relaying operation.",
+				"database does not contain latest processed block data and startBlockHeight is unset. Please provide a non-zero startBlockHeight in the configuration.",
 				zap.String("blockchainID", r.sourceBlockchainID.String()),
-				zap.Error(err),
 			)
+			return 0, errors.New("database does not contain latest processed block data and startBlockHeight is unset.")
 		}
-		return nil
+		return startBlockHeight, nil
+	} else if err != nil {
+		// Otherwise, we've encountered an unknown database error
+		r.logger.Warn(
+			"failed to get latest block from database",
+			zap.String("blockchainID", r.sourceBlockchainID.String()),
+			zap.Error(err),
+		)
+		return 0, err
 	}
-	if errors.Is(err, database.ErrChainNotFound) || errors.Is(err, database.ErrKeyNotFound) {
-		// Otherwise, latestProcessedBlock is nil, so we instead store the latest block height.
+
+	// If the database does contain the latest processed block data for the chain,
+	// use the max of the latest processed block and the configured start block height (if it was provided)
+	latestProcessedBlock, success := new(big.Int).SetString(string(latestProcessedBlockData), 10)
+	if !success {
+		r.logger.Error("failed to convert latest block to big.Int", zap.Error(err))
+		return 0, err
+	}
+	if startBlockHeight == 0 || latestProcessedBlock.Uint64() > startBlockHeight {
```

Comment:
Sanity checking here, is the `startBlockHeight == 0` condition necessary? I think `latestProcessedBlock.Uint64() > startBlockHeight` is sufficient.

It would also be good to add unit tests for this function.

Reply:
That's correct. Simplified this condition and added a unit test